### PR TITLE
(Reload Users 1/3) Add javasscript file to mod_form to refresh user list with AJAX

### PR DESCRIPTION
**Pull request by octaviorequena**
This PR includes a javascript file to the
activity moodle form in order to refresh the list
of users when the accredible group id is changed.

### How to test it locally
This PR add a new web service to the plugin; for this
to work we need to bump the plugin version so moodle
load this new service.

### Example
https://user-images.githubusercontent.com/91325424/166421817-586b0f36-2f33-4613-be46-189e7ed972a3.mov

### CODE REVIEW

File: classes/external/form_helper.php
Comment (kei178) [inline classes/external/form_helper.php:77]: Question - Wouldn't it be better to run `self::validate_context` first "before operating on any data in an external function" as written in the document? https://docs.moodle.org/dev/External_functions_API

File: classes/helpers/user_helper.php
Comment (kei178) [inline classes/helpers/user_helper.php:30]: I think it would be better to not add a new subfolder like "helpers" as somebody is saying:

> If you are going to use a subfolder, then, to avoid clashes with future core Moodle class namespaces (e.g. the 'task' and 'event' subfolders), you should create your subfolders inside 'local', e.g. myplugin/classes/local/blah - after that, you're free to do what you want (and note that the use of 'local' is not enforced, just encouraged).

https://moodle.org/mod/forum/discuss.php?d=365379

As we're keeping our plugin local logic within `classes/local`, how about naming this class just `classes/local/users.php`?

File: amd/src/userlist_updater.js
Comment (kei178) [inline amd/src/userlist_updater.js:83]: Question - Did you find any examples that append HTML elements like this to `mod_form` with jQuery? I was a bit worried about adding them without the built-in function `$mform->addElement`.

File: classes/helpers/user_helper.php
Comment (kei178) [inline classes/helpers/user_helper.php:62]: Subjective - I would make this function receive `$enrolledusers` instead of `$context` because this BE logic would be more reusable without knowing the FE context and the naming of this function will be simpler.

**Review by kei178 — CHANGES_REQUESTED** [review]
@octaviorequena - I left several comments. Please let me know if anything is unclear.

File: classes/external/form_helper.php
Comment (naveenbalachandar) [inline classes/external/form_helper.php:74]: It would be better to add tests for `form_helper` as well.

File: amd/src/userlist_updater.js
Comment (octaviorequena) [reply amd/src/userlist_updater.js:83]: I found this [plugin](https://github.com/moodleou/moodle-filter_embedquestion/blob/main/amd/src/questionid_choice_updater.js) which I use as an example. It doesn't add new inputs as we are doing, but adding new options to a select.

I thought the same as you, that maybe we could generate the new inputs with `addElement` and then append them with the javascript, but I didn't know exactly how to do that.

File: classes/external/form_helper.php
Comment (octaviorequena) [reply classes/external/form_helper.php:55]: When building external web services, we need to add two functions apart of the main one: `FUNCTION_parameters` and `FUNCTION_returns`. Those functions basically describe the parameters the main function will receive and the schema of the response retunred.

This is described in the [Moodle Docs](https://docs.moodle.org/dev/Adding_a_web_service_to_a_plugin#Write_the_external_function_descriptions)

File: classes/external/form_helper.php
Comment (octaviorequena) [reply classes/external/form_helper.php:74]: We have a problem adding a test for this method because we cannot send a mock api client to stub the request to accredible. That's why we are using the `load_users_with_credentials_from_course_context` method which can be tested without problems.

File: classes/external/form_helper.php
Comment (octaviorequena) [reply classes/external/form_helper.php:77]: In order to load the context to validate it we need to use the parameter `$courseid` and I think we need to validate and sanitise the parameters before using them, that's the reason I didn't run the validation at the start of the method.

File: mod_form.php
Comment (octaviorequena) [reply mod_form.php:224]: That's a good point, I will hide it when creating the activity and I will add a text that say something like "You need to choose an Accredible group in order to see the list of users"

File: classes/external/form_helper.php
Comment (kei178) [reply classes/external/form_helper.php:77]: That makes sense. Thanks for the explanation.

File: amd/src/userlist_updater.js
Comment (kei178) [reply amd/src/userlist_updater.js:83]: If we make sure how `$mform->addElement` works with other themes, such as using a different theme won't affect expected css styles, I would be fine with the current solution for now. We should definitely consider using the same logic to display the same HTML in a separate ticket.

Comment (kei178) [reply amd/src/userlist_updater.js:83]: I investigated how `$mform->addElement` works. I haven't completely understood it yet but I believe `$mfrom->addElement` which calls `HTML_QuickForm::addElement()` ([here](https://pear.php.net/manual/en/package.html.html-quickform.html-quickform.addelement.php)) is using mustache templates. For example:
- advcheckbox: https://github.com/moodle/moodle/blob/1a9bee69e60f687bcaf4db0569665f4186c515a4/lib/form/templates/element-advcheckbox.mustache
- static: https://github.com/moodle/moodle/blob/511a87f5fc357f18a4c53911f6e6c7f7b526246e/lib/form/templates/element-static.mustache

As you can see from the history of the files, the output HTML will change when templates are updated so we should use them in `amd/src/userlist_updater.js` as well instead of inserting static HTML.

It looks like mustache templates can be rendered from JavaScript so it would not be very difficult to replace it with the templates.
- https://docs.moodle.org/dev/Templates#Rendering_in_JavaScript
- https://docs.moodle.org/dev/User:Damyon_Wiese/Templates

Comment (kei178) [reply amd/src/userlist_updater.js:83]: We will fix this point with another ticket NTGR-679. Until then, we won't merge this PR to `develop`.

File: classes/helpers/user_helper.php
Comment (octaviorequena) [reply classes/helpers/user_helper.php]: I have moved this class to `classes/local/users` as you recommend. Thanks for the explanation!

File: classes/helpers/user_helper.php
Comment (octaviorequena) [reply classes/helpers/user_helper.php]: I have updated the method and the tests to receive the users instead of the context.

File: mod_form.php
Comment (octaviorequena) [reply mod_form.php:227]: I have added a warning message and hide the user list when there's no group chosen in the select.

File: classes/local/users.php
Comment (kei178) [inline classes/local/users.php:59]: Subjective - This function returns users with credentials as a result so `get_users_with_credentials` would sound more straightforward to me.

File: classes/local/users.php
Comment (kei178) [inline classes/local/users.php:72]: Nested loops are not time efficient as O(n^2) so I'd avoid it here writing something per below (I haven't tested it):

```php
        $users = array();
        if (!$enrolledusers) {
            return $users;
        }

        $certificatesmemo = array();
        if ($groupid) {
            $credentialsclient = new credentials($this->apirest);
            $certificates = $credentialsclient->get_credentials($groupid);
            foreach ($certificates as $certificate) {
                if (isset($certificate->url)) {
                    $credentialurl = $certificate->url;
                } else {
                    $credentialurl = 'https://www.credential.net/' . $certificate->id;
                }
                $certificatesmemo[$certificate->recipient->email] = array(
                    'credentialid' => $certificate->id,
                    'credentialurl' => $credentialurl
                );
            }
        }

        foreach ($enrolledusers as $user) {
            $certificate = $certificatesmemo[strtolower($user->email)];
            $credentialurl = isset($certificate) ? $certificate['credentialurl'] : null;
            $credentialid = isset($certificate) ? $certificate['credentialid'] : null;
            $user = array(
                'id'             => $user->id,
                'email'          => $user->email,
                'name'           => $user->firstname . ' ' . $user->lastname,
                'credential_url' => $credentialurl,
                'credential_id'  => $credentialid
            );
            array_push($users, $user);
        }
        return $users;
```

And, if it's not too much additional work, I'd reuse this function for `mod_form.php` too.

**Review by kei178 — COMMENTED** [review]
@octaviorequena - I left additional comments on `local/users.php`.

**Review by octaviorequena — COMMENTED** [review]
@naveenbalachandar @kei178 - I have pushed the latest changes which include:

- Refactor new class `Users` and use method in mod_form to improve performance to avoid fetching the credentials multiple times.
- Hide user list when adding the Accredible activity and display it only when a group has been selected.

File: tests/local/mod_accredible_users_test.php
Comment (octaviorequena) [reply tests/local/mod_accredible_users_test.php:141]: I've added a test for that case.

File: classes/local/users.php
Comment (kei178) [inline classes/local/users.php:82]: I'd move `foreach ($certificates as $certificate) { ... }` to outside of `try { ... } catch { ... } ` because I think we  want to catch errors only from `$credentialsclient->get_credentials($groupid)`.

**Review by kei178 — APPROVED** [review]
@octaviorequena Looks good, nice work👍  I left one minor comment.

**Review by naveenbalachandar — APPROVED** [review]
Nice work!

**Pull request merged**
